website/www/site/content/en/documentation/sdks/yaml-testing.md

```diff
@@ -0,0 +1,292 @@
+    --yaml_pipeline_file=wordcount.yaml \
+    --tests
+```
+
+Alternatively, the a `tests:` block may be placed in a separate file and be
+validated by running
+
+```
+python -m apache_beam.yaml.main \
+    --yaml_pipeline_file=wordcount.yaml \
+    --tests \
+    --test_suite=test_file.yaml
+```
+
+Neither the actual Read transform nor Write transform from the original
+pipelines are executed when running the test, but all intermediate transforms
+are.  For hermeticity, we require that all inputs (with the exception of
+`Create` that are needed to compute the expected outputs are explicitly mocked;
+to explicitly allow a sources to be executed as part of a test their names or
+types can be enumerated in an `allowed_sources` attribute of the test
+specification.
+
+
+## Pipeline fragment tests
+
+One can also tests a portion of a pipeline using the `mock_inputs` and
+`expected_outputs` section of a test, for example
+
+```
+tests:
+- name: TestSplittingWithPunctuation
+  mock_inputs:
+    - name: Split words
+      elements:
+        - line: "lots-of-words"
+        - line: "...and more"
+  expected_outputs:
+    - name: Explode
+      elements:
+        - word: lots
+          value: 1
+        - word: of
+          value: 1
+        - word: words
+          value: 1
+        - word: and
+          value: 1
+        - word: more
+          value: 1
+
+- name: TestCombineAndFormat
+  mock_inputs:
+    - name: Count words
+      elements:
+        - word: more
+          value: 1
+        - word: and
+          value: 1
+        - word: more
+          value: 1
+  expected_outputs:
+    - name: Format output
+      elements:
+        - output: "more: 2"
+        - output: "and: 1"
+```
+
+As before, each test only executes the portion of the pipeline between the
+mock inputs and expected outputs.  Note that the named transform in a
+`mock_inputs` specification *is* executed, while the named transform of a
+`mock_oupputs` specification is not.
+Similarly, the named transform of a `expected_inputs` specification is *not*
+executed, while the named transform of an `expected_outputs` necessarily is.
+
+
+## Automatically generating tests.
+
+In an effort to make tests as easy to write and maintain as possible,
+Beam YAML provides utilities to compute the expected outputs for your tests.
+Running
+
+```
+python -m apache_beam.yaml.main \
+    --yaml_pipeline_file=wordcount.yaml \
+    --tests \
+    [--test_suite=...] \
+    --fix_tests
+```
+
+will update any existing `expected_input` and `expected_output` blocks of your
+pipeline to contain the actual values computed during the test.
+(Of course, it is on any user of this flag to verify that the produced values
+are as expected.)
+This can be useful in authoring tests as well--one can simply specify a
+nonsensical or empty elements block in the expectation and the `--fix_tests`
+flag will populate it for you.
+
+Beam YAML also has a `--create_test` flag which can be used to create an
```

Comment:
I think it would actually be great to lead with this in this section. I expect that most users would use this to get a suite set up, then they could add more tests.

It is also worth mentioning that you need a test-suite flag when using this.

Reply:
Done.